jni/src/Foreign/JNI.hs

```diff
@@ -470,52 +543,82 @@ getStaticMethodID
   -> JNI.String -- ^ JNI signature
   -> IO JMethodID
 getStaticMethodID cls methodname sig = withJNIEnv $ \env ->
+    withJ cls $ \clsp ->
     throwIfException env $
     JNI.withString methodname $ \methodnamep ->
     JNI.withString sig $ \sigp ->
     [CU.exp| jmethodID {
       (*$(JNIEnv *env))->GetStaticMethodID($(JNIEnv *env),
-                                           $(jclass cls),
+                                           $(jclass clsp),
                                            $(char *methodnamep),
                                            $(char *sigp)) } |]
 
 getObjectClass :: Coercible o (J ty) => o -> IO JClass
 getObjectClass (coerce -> upcast -> obj) = withJNIEnv $ \env ->
-    [CU.exp| jclass {
+    withJ obj $ \objp ->
+    newJ =<< [CU.exp| jclass {
       (*$(JNIEnv *env))->GetObjectClass($(JNIEnv *env),
-                                        $(jobject obj)) } |]
+                                        $(jobject objp)) } |]
 
+-- | Creates a global reference to the object referred to by
+-- the given reference.
+--
+-- Arranges for a finalizer to call 'deleteGlobalRef' when the
+-- global reference is no longer reachable on the Haskell side.
 newGlobalRef :: Coercible o (J ty) => o -> IO o
 newGlobalRef (coerce -> upcast -> obj) = withJNIEnv $ \env ->
```

Comment:
Sorry - old comment while I was trying make sense of the difference between `newGlobalRef`, `newJ`, `newGlobalJ`. Please disregard.